docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,215 @@
+about exactly what such metadata looks like, instead we should build a single
+API for applications to use metadata associated with a project. The schemas and
+interpretations of these data types then become composable at the application
+layer.
+
+## Design Goals
+
+- Local first
+- Extensible. It should be possible to extend Radicle with new collaborative
+  data types without changing the protocol version
+- Interoperable. There should be a straightforward API for other tools to
+  interact with collaborative objects
+
+The local first goal matches Radicles peer to peer design and philosophy. You
+should not need to be online to modify collaborative objects and no one should
+be able to stop you from changing data that is on your computer. That said,
+collaborative objects are _collaborative_, users need to be able to receive
+changes to the same data from other people and this raises the problem of how
+to merge those changes. I see two options here:
+
+- CRDTs, data structures which allow conflict free merges
+- Application level merges. Much as Git requires user action to merge
+  conflicting changes, we could expose the merge points of collaborative
+  objects to the user to resolve.
+
+The latter here is undesirable for a lot of reasons but most relevant is that
+it conflicts with extensibility and interoperability. If we want to add new
+data types without changing the protocol version then the only general approach
+for application level merges is to ask the user to directly merge the
+underlying data structures. Requiring users to understand the underlying
+representations of the data they are working with is awful UX which means that
+in order to provide a good UX tools would need to provide a UI for users to do
+a three way merge of the data they are working on. This contradicts the
+interoperability goal.
+
+This leaves us with CRDTs. For the purposes of this RFC I am going to assume
+that we will use the [Automerge](https://github.com/automerge/automerge) CRDT
+implementation. Automerge is capable of merging arbitrary JSON objects (the
+data model is actually richer than JSON, with types for byte arrays, various
+precision integers and floats etc.). An alternative approach would be to write
+a custom CRDT for each data type we want to replicate, see [##Alternative
+Approaches] for a discussion of this design.
+
+## Message storage and transmission
+
+The property of CRDTs that we care about is strong eventual consistency: any
+two nodes which have received the same set of messages will merge to the same
+state. A major part of the design of collaborative objects therefore is how we
+store and transmit these messages. Messages in this context are just binary
+blobs which need to be passed to the CRDT implementation in any order.
+
+We add a new entry to the `signed_refs` of a project at
+`refs/namespaces/<project>/remotes/refs/remotes/<peer>/rad/collaborative-objects`.
```

Comment:
Yeah hashing the schema is the approach taken by the Avro Schema Registry. It works very nicely if you don't have forks in the schema evolution. In a distributed setting it's harder to see how this would work, although if we only allowed schemas to be updated via the `ProjectPayload` then it might be possible.

Reply:
It might be useful for easily detecting when you're on the same schema. Could be interesting to think about it via a merkle-chain where you could detect which side is in the past, or as you said if it's a fork. Purely spit-balling here :)